src/backtracking/knight_tour.rs

```diff
@@ -0,0 +1,206 @@
+    ///
+    /// A new KnightTour instance.
+    pub fn new(size: usize) -> Self {
+        let board = vec![vec![0; size]; size];
+        let moves = [
+            (2, 1),
+            (1, 2),
+            (-1, 2),
+            (-2, 1),
+            (-2, -1),
+            (-1, -2),
+            (1, -2),
+            (2, -1),
+        ];
+        KnightTour {
+            board_size: size,
+            board,
+            moves,
+        }
+    }
+
+    /// Checks if the given position is safe to move to.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The x-coordinate of the position.
+    /// * `y` - The y-coordinate of the position.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether the position is safe to move to.
+    fn is_safe(&self, x: i32, y: i32) -> bool {
+        x >= 0
+            && y >= 0
+            && x < self.board_size as i32
+            && y < self.board_size as i32
+            && self.board[x as usize][y as usize] == 0
+    }
+
+    /// Recursively solves the Knight's Tour problem.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The current x-coordinate of the knight.
+    /// * `y` - The current y-coordinate of the knight.
+    /// * `move_count` - The current move count.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: i32) -> bool {
+        if move_count == self.board_size as i32 * self.board_size as i32 {
+            return true;
+        }
+
+        for i in 0..8 {
+            let next_x = x + self.moves[i].0;
+            let next_y = y + self.moves[i].1;
+
+            if self.is_safe(next_x, next_y) {
+                self.board[next_x as usize][next_y as usize] = move_count + 1;
+
+                if self.solve_tour(next_x, next_y, move_count + 1) {
+                    return true;
+                }
+                // Backtrack
+                self.board[next_x as usize][next_y as usize] = 0;
+            }
+        }
+
+        false
+    }
+
+    /// Finds the Knight's Tour starting from the specified position.
+    ///
+    /// # Arguments
+    ///
+    /// * `start_x` - The x-coordinate of the starting position.
+    /// * `start_y` - The y-coordinate of the starting position.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing a boolean indicating whether a tour was found and the tour matrix if found.
+    pub fn find_tour(&mut self, start_x: usize, start_y: usize) -> (bool, Vec<Vec<i32>>) {
+        if !self.is_safe(start_x as i32, start_y as i32) {
+            println!("Invalid starting position");
+            return (false, vec![]);
+        }
+
+        self.board[start_x][start_y] = 1;
+
+        if !self.solve_tour(start_x as i32, start_y as i32, 1) {
+            println!("No solution exists");
+            return (false, vec![]);
+        }
+
+        println!("Knight's Tour:");
+        (true, self.board.clone())
+    }
+}
```

Comment:
What do you thing about having just one public function? Its signature should look something like that

```rust
pub fn find_tour(size_x: usize, size_y: usize, start_x: usize, start_y: usize) -> (bool, Vec<Vec<usize>>)
```

In my opinion nothing else should be available/visible for the external user.

Reply:
- Yeah, I think it is a reasonable design choice, especially if we want to encapsulate the implementation details of Knight's Tour and expose only necessary functionalities for user.
- By having a single public function, we can hide the internal details of the `KnightTour` struct and its methods, providing a simpler and cleaner interface for users.
- By providing the necessary parameters `(size_x, size_y, start_x, start_y)` for the chessboard size and starting position, users have the flexibility to customize their Knight's Tour problem without needing to understand the internal workings of the implementation.

Reply:
I will apply your suggestion in the next changes (commit).

Reply:
I will change the name of the public function from `find_tour` to `find_knight_tour` to make the function more descriptive.